### PR TITLE
cilium/cmd: Add cilium bpf ipcache get

### LINKED ISSUE #4880: Add `cilium bpf ipcache get <ip>`
It would be useful to be able to run something like the below to resolve which identity that the ipcache will determine for a given IP address:

```
# cilium bpf ipcache get 10.0.2.15
10.0.2.15 maps to identity 1
```

Tasks:
* [ ] After dumping the IPcache entries from the datapath, insert them into an LPM representation
* [ ] Parse the user-specified IP
* [ ] Perform a lookup in the LPM using the specified IP

### PR DESCRIPTION
**Pull request by iffyio**
This commit adds a bpf ipcache command to resolve the
identity, within the ipcache, for an ip address.

Example:
```
vagrant@runtime1:~/$ sudo cilium bpf ipcache get 10.0.2.15
10.0.2.15 maps to identity 1
```

* Add cmd/bpf_ipcache_get.go to implement command along with test
* Vendor github.com/hashicorp/go-immutable-radix This package
  is leveraged to implement a longest prefix match algorithm
* Update Documentation/cmdref with new command

Fixes #4880

Happy to change anything that isn't right!

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5665)

### CODE REVIEW

File: cilium/cmd/bpf_ipcache_get.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+		common.RequireRootPrivilege("cilium bpf ipcache get")
+
+		if len(args) < 1 || args[0] == "" {
+			Usagef(cmd, "No ip provided. "+usage)
+		}
+
+		bpfIPCacheList := dumpIPCache()
+
+		if len(bpfIPCacheList) == 0 {
+			fmt.Fprintf(os.Stderr, "No entries found.\n")
```
Comment (vadorovsky) [inline cilium/cmd/bpf_ipcache_get.go]: I think we should exit the program here. Am I wrong?

```
fmt.Fprintf(os.Stderr, "No entries found.\n")
os.Exit(1)
```

File: cilium/cmd/bpf_ipcache_get.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		ip := net.ParseIP(arg)
+		if ip == nil {
+			Usagef(cmd, "Invalid ip address. "+usage)
+		}
+
+		value, exists := getLPMValue(ip, bpfIPCacheList)
+
+		if !exists {
+			fmt.Printf("%s does not map to any identity\n", arg)
+			return
```
Comment (vadorovsky) [inline cilium/cmd/bpf_ipcache_get.go]: I think this message should also go to stderr. And instead of doing `return` (which will return status 0), we should exit with status 1.

```
fmt.Fprintf(os.Stderr, "%s does not map to any identity\n", arg)
os.Exit(1)
```

**Review by vadorovsky — CHANGES_REQUESTED** [review]
Great work and very useful command! I have only few minor comments.

File: cilium/cmd/bpf_ipcache_get.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+		}
+
+		v := value.([]string)
+		if len(v) == 0 {
+			fmt.Printf("Unable to retrieve identity for LPM entry %s\n", arg)
+			return
```
Comment (vadorovsky) [inline cilium/cmd/bpf_ipcache_get.go]: Same here.

```
fmt.Fprintf(os.Stderr, "Unable to retrieve identity for LPM entry %s\n", arg)
os.Exit(1)
```

**Comment by tgraf** [comment]
test-me-please

**Comment by iffyio** [comment]
Updated :) Thanks for catching this!

**Review by vadorovsky — APPROVED** [review]

**Comment by vadorovsky** [comment]
test-me-please

File: cilium/cmd/bpf_ipcache_get_test.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
+package cmd
```
Comment (tgraf) [inline cilium/cmd/bpf_ipcache_get_test.go:15]: Copyright header

**Review by tgraf — CHANGES_REQUESTED** [review]
Awesome!

File: cilium/cmd/bpf_ipcache_get_test.go (hunk 4 above)
Comment (iffyio) [reply cilium/cmd/bpf_ipcache_get_test.go:15]: Fixed :sweat_smile:

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
test-me-please

File: cilium/cmd/bpf_ipcache_get.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+		ip := net.ParseIP(arg)
+		if ip == nil {
+			Usagef(cmd, "Invalid ip address. "+usage)
+		}
+
+		bpfIPCacheList := dumpIPCache()
+
+		if len(bpfIPCacheList) == 0 {
+			fmt.Fprintf(os.Stderr, "No entries found.\n")
+			os.Exit(1)
```
Comment (aanm) [inline cilium/cmd/bpf_ipcache_get.go]: Shouldn't this be exit(0)?

**Comment by iffyio** [comment]
trying to figure out the endpoint failures on the CI, it seems unrelated?

File: cilium/cmd/bpf_ipcache_get.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+func dumpIPCache() map[string][]string {
+	bpfIPCacheList := make(map[string][]string)
+
+	if err := ipcache.IPCache.Dump(bpfIPCacheList); err != nil {
+		Fatalf("unable to dump IPCache: %s\n", err)
+	}
+
+	return bpfIPCacheList
```
Comment (joestringer) [inline cilium/cmd/bpf_ipcache_get.go]: As a matter of style, it's good to avoid naming a map as a list as they're different underlying structure types. (Same for everywhere that is passing this map around and naming it `bpfIPCacheList`.)

File: cilium/cmd/bpf_ipcache_get_test.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	}
+
+	if isIPV4(ip) {
+		ip = ip.To4()
+	}
+
+	return ip
+}
+
+func toBits(bytes []byte) []byte {
```
Comment (aanm) [inline cilium/cmd/bpf_ipcache_get_test.go:106]: It seems this function has the same logic of the `GetPrefix` function. Instead of using this function can you hardcode the expected bits in the tests?

**Review by aanm — CHANGES_REQUESTED** [review]

**Review by joestringer — CHANGES_REQUESTED** [review]
I have a minor style suggestion below but otherwise looks good. Thanks!

File: cilium/cmd/bpf_ipcache_get.go (hunk 5 above)
Comment (joestringer) [reply cilium/cmd/bpf_ipcache_get.go]: No, because the user is explicitly looking for a particular IP in the trie.

**Comment by joestringer** [comment]
@iffyio I looked into the failure. From this page, you can download the zip:

https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/6685/testReport/k8s-1/11/K8sUpdates_Tests_upgrade_and_downgrade_from_a_Cilium_stable_image_to_master/

[83db7c8f_K8sUpdates_Tests_upgrade_and_downgrade_from_a_Cilium_stable_image_to_master.zip](https://github.com/cilium/cilium/files/2412408/83db7c8f_K8sUpdates_Tests_upgrade_and_downgrade_from_a_Cilium_stable_image_to_master.zip)

Looking into the logs file `pod-kube-system-cilium-cr7r2-cilium-agent.log`, there is this warning which is likely causing the failure:

```
2018-09-24T17:10:53.157338043Z level=warning msg="Command execution failed" cmd="/var/lib/cilium/bpf/join_ep.sh /var/lib/cilium/bpf /var/run/cilium/state /var/run/cilium/state/53014_next lxc98bad false 53014" containerID=98badf46c8 endpointID=53014 error="signal: killed" ipv4=10.10.0.63 ipv6="f00d::a0a:0:0:cf16" k8sPodName=/ policyRevision=0
2018-09-24T17:10:53.157401067Z level=warning msg="Join EP id=/var/run/cilium/state/53014_next ifname=lxc98bad" subsys=endpoint
2018-09-24T17:10:53.157405102Z level=warning msg="'probe' is not a recognized processor for this target (ignoring processor)" subsys=endpoint
2018-09-24T17:10:53.157407094Z level=warning msg="'probe' is not a recognized processor for this target (ignoring processor)" subsys=endpoint
2018-09-24T17:10:53.157409143Z level=info msg="Recompiled endpoint BPF program" BPFCompilationTime=842.786878ms containerID=98badf46c8 endpointID=53014 error="error: \"signal: killed\" command output: \"Join EP id=/var/run/cilium/state/53014_next ifname=lxc98bad\\n'probe' is not a recognized processor for this target (ignoring processor)\\n'probe' is not a recognized processor for this target (ignoring processor)\\n\"" ipv4=10.10.0.63 ipv6="f00d::a0a:0:0:cf16" k8sPodName=/ policyRevision=0
2018-09-24T17:10:53.157412089Z level=info msg="Regeneration of BPF program has completed" buildDuration=863.122152ms containerID=98badf46c8 endpointID=53014 ipv4=10.10.0.63 ipv6="f00d::a0a:0:0:cf16" k8sPodName=/ policyRevision=0
2018-09-24T17:10:53.157415298Z level=warning msg="Generating BPF for endpoint failed, keeping stale directory." containerID=98badf46c8 endpointID=53014 file-path=53014_next_fail ipv4=10.10.0.63 ipv6="f00d::a0a:0:0:cf16" k8sPodName=/ policyRevision=0
2018-09-24T17:10:53.157448355Z level=debug msg="Completed endpoint regeneration with no pending regeneration requests" code=OK containerID=98badf46c8 endpointID=53014 endpointState=ready ipv4=10.10.0.63 ipv6="f00d::a0a:0:0:cf16" k8sPodName=/ policyRevision=0 type=0
2018-09-24T17:10:53.157562989Z level=warning msg="Regeneration of endpoint program failed" containerID=98badf46c8 endpointID=53014 error="error: \"signal: killed\" command output: \"Join EP id=/var/run/cilium/state/53014_next ifname=lxc98bad\\n'probe' is not a recognized processor for this target (ignoring processor)\\n'probe' is not a recognized processor for this target (ignoring processor)\\n\"" ipv4=10.10.0.63 ipv6="f00d::a0a:0:0:cf16" k8sPodName=/ policyRevision=0
2018-09-24T17:10:53.157569057Z level=debug msg="Error regenerating endpoint: error: \"signal: killed\" command output: \"Join EP id=/var/run/cilium/state/53014_next ifname=lxc98bad\\n'probe' is not a recognized processor for this target (ignoring processor)\\n'probe' is not a recognized processor for this target (ignoring processor)\\n\"" code=Failure containerID=98badf46c8 endpointID=53014 endpointState=ready ipv4=10.10.0.63 ipv6="f00d::a0a:0:0:cf16" k8sPodName=/ policyRevision=0 type=200
2018-09-24T17:10:53.157607202Z level=warning msg="Failed while regenerating endpoint" endpointID=53014 subsys=daemon
```

Basically it seems that for some reason the compiler was killed when it attempted to recompile the programs for one of the endpoints.

Given that your PR is entirely on the CLI side, this is unrelated and safe for you to ignore.

File: cilium/cmd/bpf_ipcache_get_test.go (hunk 7 above)
Comment (iffyio) [reply cilium/cmd/bpf_ipcache_get_test.go:106]: yes, both functions are quite similar, but the alternative as you suggest would require hard-coding byte arrays of length 32 and 128 for v4 and v6 addresses respectively. is this preferred?

File: cilium/cmd/bpf_ipcache_get.go (hunk 6 above)
Comment (iffyio) [reply cilium/cmd/bpf_ipcache_get.go]: Fixed :)

**Comment by joestringer** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

File: cilium/cmd/bpf_ipcache_get_test.go (hunk 7 above)
Comment (tgraf) [reply cilium/cmd/bpf_ipcache_get_test.go:106]: I think this is fine to get started.
